sql: accommodate no zone config for table/index on schema changer

vivekmenezes opened the pull request:
The schema changer process runs asynchronously and can
therefore sometimes not see a zone config that has already been
dropped. It should be resilient to such failures.

fixes #35988 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35991)
<!-- Reviewable:end -->

vivekmenezes commented:
I'll look into adding a unittest

vivekmenezes committed:
sql: accommodate no zone config for table/index on schema changer

The schema changer process runs asynchronously and can
therefore sometimes not see a zone config that has already been
dropped. It should be resilient to such failures.

Release note: None

knz reviewed (commented):
LGTM. I think you can write a test by having a SQL query delete the default zone from `system.zone`?

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-:-LaRtEb7Ekozmi4rrtkI:bivk2iv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @vivekmenezes)

___
*[pkg/sql/schema_changer.go, line 1635 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-LaRsaJFBP8NcAw9-ItK:-LaRsaJFBP8NcAw9-ItL:bg1dbv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae0cace4934e12fce2a076c0a7741fea21275ac/pkg/sql/schema_changer.go#L1635)):*
> ```Go
> 						zoneCfg, placeholder, _, err := ZoneConfigHook(cfg, uint32(id))
> 						if err != nil {
> 							log.Errorf(ctx, "zone config for desc: %d, err = %s", id, err)
> ```

please use `%+v` for errors (or at least '%v')
___
*[pkg/sql/schema_changer.go, line 1735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-LaRsdql92I2Ndl93Z0A:-LaRsdql92I2Ndl93Z0B:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae0cace4934e12fce2a076c0a7741fea21275ac/pkg/sql/schema_changer.go#L1735)):*
> ```Go
> 							zoneCfg, placeholder, _, err := ZoneConfigHook(cfg, uint32(table.ID))
> 							if err != nil {
> 								log.Errorf(ctx, "zone config for desc: %d, err = %s", table.ID, err)
> ```

ditto
___
*[pkg/sql/schema_changer.go, line 1787 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-LaRt1vb9jTQkpDnBCNA:-LaRt1vb9jTQkpDnBCNB:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae0cace4934e12fce2a076c0a7741fea21275ac/pkg/sql/schema_changer.go#L1787)):*
> ```Go
> 								zoneCfg, _, _, err := ZoneConfigHook(cfg, uint32(table.ID))
> 								if err != nil {
> 									log.Errorf(ctx, "zone config for desc: %d, err: %s", table.ID, err)
> ```

ditto


<!-- Sent from Reviewable.io -->

dt commented on pkg/sql/schema_changer.go:
```diff
@@ -1728,22 +1732,25 @@ func (s *SchemaChangeManager) Start(stopper *stop.Stopper) {
-								if subzone := placeholder.GetSubzone(uint32(m.IndexID), ""); subzone != nil {
-									ttlSeconds = subzone.Config.GC.TTLSeconds
+								// Initialize deadline without a zone config in case it's not present.
+								deadline := m.DropTime + int64(delay)
```
should we use default zone ttl instead of dropping immediately?

vivekmenezes reviewed (commented):
I've been having problems reproducing the problem by deleting the zone config. This is all a mystery to me. In any case I think the change is more conservative now in that in the absence of a zone config it chooses the default one for TTL. I'm going ahead and merging it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-:-LaT0qIWCcnLsE5ihlUg:b-59t3g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @knz)

___
*[pkg/sql/schema_changer.go, line 1635 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-LaRsaJFBP8NcAw9-ItK:-LaT-tbi6RGmS6qe25NI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae0cace4934e12fce2a076c0a7741fea21275ac/pkg/sql/schema_changer.go#L1635)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

please use `%+v` for errors (or at least '%v')
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer.go, line 1735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-LaRsdql92I2Ndl93Z0A:-LaT-uh0E_aUsrydQYG3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae0cace4934e12fce2a076c0a7741fea21275ac/pkg/sql/schema_changer.go#L1735)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer.go, line 1745 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-LaQz7O9fWNDRIRotk0e-r1-1745:-LaT-vNjAYGF2AOaLydW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae0cace4934e12fce2a076c0a7741fea21275ac/pkg/sql/schema_changer.go#L1745)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

should we use default zone ttl instead of dropping immediately?
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer.go, line 1787 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-LaRt1vb9jTQkpDnBCNA:-LaT-w8m-kkel1GoTP3b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae0cace4934e12fce2a076c0a7741fea21275ac/pkg/sql/schema_changer.go#L1787)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I'm really fighting how to trigger a errNoZoneConfigApplies error. I do see a TODO that nikhil left 
https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/show_zone_config.go#L111

I'm going to go ahead and merge this.

knz commented:
Have you considered sending a blind KV del on the zone table, instead of going through SQL? That's how I was testing this when I was implementing SQL configs.

vivekmenezes commented:
bors r+

vivekmenezes commented:
bors r-

vivekmenezes reviewed (commented):
@knz I don't believe it should matter. Do you have an example of this checked in?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35991#-:-LaWM-YK2YhmseWdHJDK:b3z5o6b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @knz)



<!-- Sent from Reviewable.io -->

knz commented:
To be honest I do not know whether it matters. I was  already happy with your PR earlier and I won't let the missing test block it :)

I was just pointing out that the KV interface is a good way to work around the SQL permission check.

vivekmenezes commented:
got it. I was not running into permission issues. I've tried dropping the zone config for both table and database and they succeed but never trigger the bug. I'm fixing this bug purely through code inspection. Thanks!

vivekmenezes commented:
bors r+
